app/controllers/concerns/account_concern.rb

```diff
@@ -104,4 +78,72 @@ def logout!
       }
     end
   end
+
+  def pre_results
+    results_in_account = fetch_results_from_account_or_logout
+    return unless logged_in?
+
+    now = Time.zone.now.to_i
+    @results_differ = criteria_keys != results_in_account.fetch("criteria_keys", [])
+    @results_saved = !@results_differ && results_in_account.fetch("timestamp", now) >= now - 10
+  end
+
+  def pre_saved_results
+    results_in_account = fetch_results_from_account_or_logout
+    redirect_to logged_out_pre_saved_results_path and return unless logged_in?
+
+    @saved_results = results_in_account.fetch("criteria_keys", [])
+  end
+
+  def pre_update_results
+    results_in_account = fetch_results_from_account_or_logout
+    redirect_to logged_out_pre_update_results_path and return unless logged_in?
+
+    @saved_results = results_in_account.fetch("criteria_keys", [])
+  end
+
+  def logged_out_pre_saved_results_path
+    transition_checker_new_session_url(redirect_path: transition_checker_saved_results_path, _ga: params[:_ga])
+  end
+
+  def logged_out_pre_update_results_path
+    transition_checker_new_session_url(redirect_path: transition_checker_save_results_confirm_path(c: criteria_keys), _ga: params[:_ga])
+  end
+
+  def fetch_results_from_account_or_logout
+    result = do_or_logout { Services.account_api.get_attributes(govuk_account_session: account_session_header, attributes: [ATTRIBUTE_NAME]) }
+    result&.dig("values", ATTRIBUTE_NAME) || {}
+  end
+
+  def fetch_email_subscription_from_account_or_logout
+    result = do_or_logout { Services.account_api.check_for_email_subscription(govuk_account_session: account_session_header) }
+    result&.fetch("has_subscription", false)
+  end
+
+  def update_email_subscription_in_account_or_logout(slug)
+    do_or_logout { Services.account_api.set_email_subscription(govuk_account_session: account_session_header, slug: slug) }
+  end
+
+  def update_answers_in_account_or_logout(new_criteria_keys)
+    do_or_logout { Services.account_api.set_attributes(govuk_account_session: account_session_header, attributes: { ATTRIBUTE_NAME => { criteria_keys: new_criteria_keys, timestamp: Time.zone.now.to_i } }) }
+  end
+
+  def do_or_logout
+    return unless account_session_header
+
+    result = yield.to_h
+    set_account_session_header(result["govuk_account_session"])
+    result
+  rescue GdsApi::HTTPUnauthorized
+    logout!
+    nil
+  end
+
+  def transition_checker_new_session_url(**params)
+    "#{Plek.find('frontend')}/sign-in?#{params.compact.to_query}"
+  end
+
+  def transition_checker_end_session_url(**params)
+    "#{Plek.find('frontend')}/sign-out?#{params.compact.to_query}"
+  end
```

Comment:
I've realised that we don't actually want `Plek.find("frontend")` *in production*, as that gives `"https://frontend.production.govuk-internal.digital"`.  But we *do* want it in development, or the links won't work.

So I think we want `Rails.env.production? ? Plek.new.website_root : Plek.find("frontend")`

Reply:
I thought all the plek env vars on prod would prevent that?
Isn't that sort of flexibility across environments the point of Plek?

Reply:
Hmm, `Plek.new.website_root` would actually work for dev, *if* you have a running router.  Which is admittedly usually not the case, but maybe it's good enough.  And it avoids the conditional here.